src/Nethermind/Nethermind.Network/ForkInfo.cs

```diff
@@ -70,5 +85,142 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
+                    return IForkInfo.ValidationResult.Valid;
+                }
+
+                // The local and remote nodes are in different forks currently, check if the
+                // remote checksum is a subset of our local forks (rule #2).
+                for (int j = 0; j < i; j++)
+                {
+                    if (Bytes.AreEqual(Forks[j].Id.ForkHash, peerId.ForkHash))
+                    {
+                        // Remote checksum is a subset, validate based on the announced next fork
+                        if (Forks[j + 1].Activation.Activation != peerId.Next)
+                        {
+                            return IForkInfo.ValidationResult.RemoteStale;
+                        }
+
+                        return IForkInfo.ValidationResult.Valid;
+                    }
+                }
+
+                // Remote chain is not a subset of our local one, check if it's a superset by
+                // any chance, signalling that we're simply out of sync (rule #3).
+                for (int j = i + 1; j < Forks.Length; j++)
+                {
+                    if (Bytes.AreEqual(Forks[j].Id.ForkHash, peerId.ForkHash))
+                    {
+                        // Yay, remote checksum is a superset, ignore upcoming forks
+                        return IForkInfo.ValidationResult.Valid;
+                    }
+                }
+                // No exact, subset or superset match. We are on differing chains, reject.
+                return IForkInfo.ValidationResult.IncompatibleOrStale;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        /// <summary>
+        /// Verify that the forkid from peer matches our forks. Code is largely copied from Geth.
+        /// </summary>
+        /// <param name="peerId"></param>
+        /// <returns></returns>
+        public IForkInfo.ValidationResult ValidateForkId2(ForkId peerId)
+        {
+            BlockHeader? head = _blockTree.Head?.Header;
+            if (head == null) return IForkInfo.ValidationResult.Valid;
+            if (!DictForks.TryGetValue(BitConverter.ToUInt32(peerId.ForkHash), out (ForkActivation Activation, ForkId Id) found))
+            {
+                return IForkInfo.ValidationResult.IncompatibleOrStale;
+            }
+            bool usingTimestamp = (found.Id.Next == 0
+                || found.Id.Next >= TimstampThreshold)
+                && (peerId.Next == 0
+                || peerId.Next >= TimstampThreshold);
+            ulong headActivation = (usingTimestamp ? head.Timestamp : (ulong)head.Number);
+
+            // my approach is to accept all peers except the ones we dont like. which is the oposite of what
+            // geth does. they reject all except ones they think are right.
+
+            if (found.Id.Next != peerId.Next)
+            {
+                if (peerId.Next == 0
```

Comment:
might be easier to read when named correctly, suggestions (not sure if 100% correct):
`peerId.Next == 0` -> `peerOnLastFork`
`found.Id.Next == 0` -> `localOnLastFork`
`found.Id.Next > 0` -> `!localOnLastFork`
`headActivation >= found.Id.Next` -> `headPastLastFork`